src/utils/post_list.js

```diff
@@ -0,0 +1,404 @@
+export function makeFilterPostsAndAddSeparators() {
+    const getPostsForIds = makeGetPostsForIds();
+
+    return createIdsSelector(
+        (state, {postIds}) => getPostsForIds(state, postIds),
+        (state, {lastViewedAt}) => lastViewedAt,
+        (state, {indicateNewMessages}) => indicateNewMessages,
+        (state) => state.entities.posts.selectedPostId,
+        getCurrentUser,
+        shouldShowJoinLeaveMessages,
+        (posts, lastViewedAt, indicateNewMessages, selectedPostId, currentUser, showJoinLeave) => {
+            if (posts.length === 0 || !currentUser) {
+                return [];
+            }
+
+            const out = [];
+
+            let lastDate = null;
+            let addedNewMessagesIndicator = false;
+
+            // Iterating through the posts from oldest to newest
+            for (let i = posts.length - 1; i >= 0; i--) {
+                const post = posts[i];
+
+                if (
+                    !post ||
+                    (post.type === Posts.POST_TYPES.EPHEMERAL_ADD_TO_CHANNEL && !selectedPostId)
+                ) {
+                    continue;
+                }
+
+                // Filter out join/leave messages if necessary
+                if (shouldFilterJoinLeavePost(post, showJoinLeave, currentUser.username)) {
+                    continue;
+                }
+
+                // Push on a date header if the last post was on a different day than the current one
+                const postDate = new Date(post.create_at);
+                postDate.setHours(0, 0, 0, 0);
+
+                if (!lastDate || lastDate.toDateString() !== postDate.toDateString()) {
+                    out.push(DATE_LINE + postDate.getTime());
+
+                    lastDate = postDate;
+                }
+
+                if (
+                    lastViewedAt &&
+                    post.create_at > lastViewedAt &&
+                    post.user_id !== currentUser.id &&
+                    !addedNewMessagesIndicator &&
+                    indicateNewMessages
+                ) {
+                    out.push(START_OF_NEW_MESSAGES);
+                    addedNewMessagesIndicator = true;
+                }
+
+                out.push(post.id);
+            }
+
+            // Flip it back to newest to oldest
+            return out.reverse();
+        }
+    );
+}
+
+export function makeCombineUserActivityPosts() {
+    return createIdsSelector(
+        (state, postIds) => postIds,
+        (state) => state.entities.posts.posts,
+        (postIds, posts) => {
+            let lastPostIsUserActivity = false;
+            let combinedCount = 0;
+
+            const out = [];
+            let changed = false;
+
+            for (let i = 0; i < postIds.length; i++) {
+                const postId = postIds[i];
+
+                if (postId === START_OF_NEW_MESSAGES || postId.startsWith(DATE_LINE)) {
+                    // Not a post, so it won't be combined
```

Comment:
Does it mean that we are combining system posts per day?  If yes, then maybe worth mentioning to PM/UX team.

Reply:
I actually would've expected the previous version to not combine posts across days, but I can understand why it would happen. That being said, it would be very difficult to combine them across days when doing it this way.

@esethna Any concerns with making this change? Currently, if you invite someone to a channel at 11:59pm and then invite someone else at 12:01am, it will combine those messages, but with these changes it'll no longer do that since there's a date separator in the way.